---
title: "Detecting Differential Expression in Single-Cell RNAseq data"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{wasserstein_singlecell}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```



The [`waddR`](waddR.html) package provides an adaptation of the semi-parametric
testing procedure based on the 2-Wasserstein distance which is specifically
tailored to identify differential distributions in single-cell RNA-seqencing
(scRNA-seq) data. 


In particular, a two-stage (TS) approach has been implemented that takes account
of the specific nature of scRNA-seq data by separately testing for differential
proportions of zero gene expression (using a logistic regression model) and
differences in non-zero gene expression (using the semi-parametric
2-Wasserstein distance-based test) between two conditions.

## Application to an example data set

The waddR package can be loaded with 
```{R}
library(waddR)
```

This an example on how to analyse the expression of single cell data in two
conditions.
As example data we will look at single cell data from decidua and blood samples
analysed by Vento-Tormo et al. 2018 
(https://doi.org/10.1038/s41586-018-0698-6).
The replicate f20 has been normalized, scaled and is available for download on
 [github](https://github.com/goncalves-lab/waddR-data/).

For this demonstration we will first load all required packages.

```{r}
suppressPackageStartupMessages({
    library("scater")
    library("SingleCellExperiment")
    library("BiocFileCache")
})
```

Download the example data

```{r}
url.base <- "https://github.com/goncalves-lab/waddR-data/blob/master/data/"
sce.blood.url <- paste0(url.base, "sce_blood.RDa?raw=true")
sce.decidua.url <- paste0(url.base, "sce_decidua.RDa?raw=true")

getCachedPath <- function(url, rname){
    bfc <- BiocFileCache() # fire up cache
    res <- bfcquery(bfc, url, field="fpath", exact=TRUE)
    if (bfccount(res) == 0L)
        cachedFilePath <- bfcadd(bfc, rname=rname, fpath=url)
    else
        cachedFilePath <- bfcpath(bfc, res[["rid"]])
    cachedFilePath
}

load(getCachedPath(sce.blood.url, "sce.blood"))
load(getCachedPath(sce.decidua.url, "sce.decidua"))
```

We have downloaded the two `SingleCellExperiment` objects `sce.blood` and
`sce.decidua`.
Next we will randomly select a small set of genes, a whole analysis would be
out of scope for this vignette.

```{r}
set.seed(24)
randgenes <- sample(rownames(sce.blood), 5000, replace=FALSE)
sce.blood <- sce.blood[randgenes, ]
sce.decidua <- sce.decidua[randgenes, ]
```


Next we prepare the input for the function `wasserstein.sc` which for now is a
dataframe or matrix object and a vector, assigning their columns to conditions.

```{r}
dat <- cbind(counts(sce.blood), counts(sce.decidua))
condition <- c(rep(1, length(colnames(sce.blood))),   # condition 1
               rep(2, length(colnames(sce.decidua)))) # condition 2
```


Before testing for differential expression in the two groups the data would
usually be normalized for cell-specific biases or technical effects.
To avoid losing actual differences in expression between cells of the same
condition, cells are also assigned to clusters and normalization only be 
performed per cluster.

Since these cells have already been preprocessed, we can run the wasserstein
test for single cell data right away.

```{r}
wsres <- wasserstein.sc(dat, condition, permnum = 10000, "TS")
head(wsres)
```



## See Also

* Fast and accurate 
[calculation of the  Wasserstein distance](wasserstein_metric.html)

* [Two-sample test](wasserstein_test.html) to check for differences between two
distributions

