---
title: "Detecting Differential Expression in Single-Cell RNAseq data"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{wasserstein_singlecell}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```



The [`waddR`](waddR.html) package provides an adaptation of the semi-parametric
testing procedure based on the 2-Wasserstein distance which is specifically
tailored to identify differential distributions in single-cell RNA-seqencing
(scRNA-seq) data. 


In particular, a two-stage (TS) approach has been implemented that takes account
of the specific nature of scRNA-seq data by separately testing for differential
proportions of zero gene expression (using a logistic regression model) and
differences in non-zero gene expression (using the semi-parametric
2-Wasserstein distance-based test) between two conditions.

## Application to an example data set

This an example on how to analyse the expression of single cell data in two
conditions.
As example data we will look at single cell data from bone marrow and umbilical
cord blood cells, which can be read be downloaded with the `HCAData` package
which returns `SummarizedExperiment` representations of the data.

(For a more detailed description of how to handle datasets from the conquer
database, see the 
[conquer R tutorial](https://github.com/markrobinsonuzh/conquer/blob/master/shiny-download/tutorial.md).)

First, we load all required libraries needed for our example.

```{r}
library("tictoc")
tic()

library(waddR)
suppressPackageStartupMessages({
    library("HCAData")
    library("ExperimentHub")
    library("SingleCellExperiment")})

```

Download the example data

```
eh <- ExperimentHub()
bonemarrow_h5densematrix <- eh[["EH2047"]]
bonemarrow_coldata <- eh[["EH2048"]]
bonemarrow_rowdata <- eh[["EH2049"]]
```

And combine into `SingleCellExperiment` representations:
```{r}
sce_bonemarrow <- HCAData("ica_bone_marrow")
sce_umbcordblood <- HCAData("ica_cord_blood")
```

Because analyzing the entire 378000 bone marrow and 384000 umbilical cord
blood cells is out of scope for this vignette, we will use random subsets of
the two groups.

```{r}
set.seed(24)
sce_bm <- sce_bonemarrow[, sample(colnames(sce_bonemarrow),
                                  100, replace=FALSE)]
set.seed(24)
sce_ucb <- sce_umbcordblood[, sample(colnames(sce_umbcordblood),
                                     100, replace=FALSE)]
```

Before testing for differential expression in the two groups we normalie
cell-specific biases with the normalize function from the `scran` package.
To avoid losing actual differences in expression between cells of the same
condition, we cluster for internal differences first and then normalize on
these clusters.

```{r}
library("scran")
#bm.clusters <- quickCluster(sce_bm, method="igraph", irlba.args=list(maxit=1000))
#sce_bm <- computeSumFactors(sce_bm, clusters=bm.clusters)
sce_bm <- computeSumFactors(sce_bm)
#ucb.clusters <- quickCluster(sce_bm, method="igraph", irlba.args=list(maxit=1000))
#sce_ucb <- computeSumFactors(sce_bm, clusters=ucb.clusters)
sce_ucb <- computeSumFactors(sce_ucb)
```

Now we will test for differential expression between the two conditions with
the `wasserstein.sc` testing procedure.
```{r}
dat <- cbind(assay(sce_bm, "counts"), assay(sce_ucb, "counts"))
conditions <- c(rep(1,100), rep(2,100))
res <- wasserstein.sc(dat, conditions, 1000, method="TS")

toc()
```

## See Also

* Fast and accurate 
[calculation of the  Wasserstein distance](wasserstein_metric.html)

* [Two-sample test](wasserstein_test.html) to check for differences between two
distributions

